protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,509 @@
+///    #     PeerId::random(),
+///    #     inner_transport,
+///    # );
+///    transport.dial(Multiaddr::empty().with(Protocol::Memory(42)));
+///    ```
+///
+/// 2. Establish relayed connections by dialing `/p2p-circuit` addresses.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_relay::v2::client;
+///    # let inner_transport = MemoryTransport::default();
+///    # let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+///    #     PeerId::random(),
+///    #     inner_transport,
+///    # );
+///    let dst_addr_via_relay = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(PeerId::random().into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit) // Signal to connect via relay and not directly.
+///        .with(Protocol::Memory(42)) // Destination address.
+///        .with(Protocol::P2p(PeerId::random().into())); // Destination peer id.
+///    transport.dial(dst_addr_via_relay).unwrap();
+///    ```
+///
+/// 3. Listen for incoming relayed connections via specific relay.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_relay::v2::client;
+///    # let inner_transport = MemoryTransport::default();
+///    # let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+///    #     PeerId::random(),
+///    #     inner_transport,
+///    # );
+///    let relay_addr = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(PeerId::random().into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit); // Signal to listen via remote relay node.
+///    transport.listen_on(relay_addr).unwrap();
+///    ```
+#[derive(Clone)]
+pub struct ClientTransport<T: Clone> {
+    to_behaviour: mpsc::Sender<TransportToBehaviourMsg>,
+
+    inner_transport: T,
+}
+
+impl<T: Clone> ClientTransport<T> {
+    /// Create a new [`ClientTransport`] by wrapping an existing [`Transport`] in a
+    /// [`ClientTransport`].
+    ///
+    ///```
+    /// # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+    /// # use libp2p_core::transport::memory::MemoryTransport;
+    /// # use libp2p_relay::v2::client;
+    /// let inner_transport = MemoryTransport::default();
+    /// let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+    ///    PeerId::random(),
+    ///    inner_transport,
+    /// );
+    ///```
+    pub(crate) fn new(t: T) -> (Self, mpsc::Receiver<TransportToBehaviourMsg>) {
+        let (to_behaviour, from_transport) = mpsc::channel(0);
+
+        let transport = ClientTransport {
+            to_behaviour,
+
+            inner_transport: t,
```

Comment:
Could you use `OrTransport` instead? It seems like that does what this is doing, so you may save some code. (https://docs.rs/libp2p/0.41.0/libp2p/core/transport/struct.OrTransport.html)

Reply:
The difference is one function call right?
This way:
```
relayTransport.dial()
  -> sees Err(TransportError::MultiaddrNotSupported); call innerTransport.dial() ...
```

OrTransport
```
orTransportDial()
  -> relayTransport.dial() -> Err(TransportError::MultiaddrNotSupported). 
  -> innerTransport.dial ...
```

Or am I missing something? If the above is true it seems like we should use OrTransport as long as relayTransport fails fast.

Reply:
Thanks for the follow up Marco!

> we should use OrTransport as long as relayTransport fails fast

:man_facepalming: I didn't consider putting the relay transport first.

08d6637732d0e792e69f0f86e26c12ca5bc3b8e7 applies your suggestion. Much better this way!